File: src/main/java/com/lcaohoanq/nocket/domain/user/IUserService.java
package com.lcaohoanq.nocket.domain.user;

import com.lcaohoanq.nocket.api.PageResponse;
import com.lcaohoanq.nocket.base.exception.DataNotFoundException;
import java.util.List;
import java.util.UUID;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.transaction.annotation.Transactional;
import com.lcaohoanq.nocket.domain.auth.AuthPort.UpdatePasswordDTO;

public interface IUserService {

    PageResponse<UserPort.UserResponse> fetchUser(Pageable pageable);

    String loginOrRegisterGoogle(String email, String name, String googleId, String avatarUrl)
        throws Exception;

    UserPort.UserResponse findUserById(UUID id) throws DataNotFoundException;

    User findUserByEmail(String email) throws DataNotFoundException;

    List<User> getAllUsers();

    User findByUsername(String username) throws DataNotFoundException;

    Page<User> findAll(String keyword, Pageable pageable) throws Exception;

    void blockOrEnable(UUID userId, Boolean active) throws DataNotFoundException;

    @Transactional
    User updateUser(UUID userId, UserPort.UpdateUserDTO updatedUserDTO) throws Exception;

    @Transactional
    User updateUserBalance(UUID userId, Long payment) throws Exception;

    void bannedUser(UUID userId) throws DataNotFoundException;

    void updatePassword(UpdatePasswordDTO updatePasswordDTO) throws Exception;

    void softDeleteUser(UUID userId) throws DataNotFoundException;

    void restoreUser(UUID userId) throws DataNotFoundException;

    void validateAccountBalance(User user, long basePrice);

    Boolean existsByEmail(String email);

    Boolean existsByPhoneNumber(String phoneNumber);

    Boolean existsById(UUID id);


}

File: server/src/main/kotlin/com/lcaohoanq/nocket/domain/user/IUserService.kt
package com.lcaohoanq.nocket.domain.user

import com.lcaohoanq.nocket.api.PageResponse
import com.lcaohoanq.nocket.base.exception.DataNotFoundException
import com.lcaohoanq.nocket.domain.auth.AuthPort.UpdatePasswordDTO
import com.lcaohoanq.nocket.domain.user.UserPort.UpdateUserDTO
import com.lcaohoanq.nocket.domain.user.UserPort.UserResponse
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.transaction.annotation.Transactional
import java.util.*

interface IUserService {
    fun fetchUser(pageable: Pageable): PageResponse<UserResponse>

    @Throws(Exception::class)
    fun loginOrRegisterGoogle(
        email: String,
        name: String,
        googleId: String,
        avatarUrl: String
    ): String

    @Throws(DataNotFoundException::class)
    fun findUserById(id: UUID): UserResponse

    @Throws(DataNotFoundException::class)
    fun findUserByEmail(email: String): User

    fun getAllUsers(): List<User>

    @Throws(DataNotFoundException::class)
    fun findByUsername(username: String): User

    @Throws(Exception::class)
    fun findAll(keyword: String, pageable: Pageable): Page<User>?

    @Throws(DataNotFoundException::class)
    fun blockOrEnable(userId: UUID, active: Boolean)

    @Transactional
    @Throws(Exception::class)
    fun updateUser(userId: UUID, updatedUserDTO: UpdateUserDTO): User

    @Throws(DataNotFoundException::class)
    fun bannedUser(userId: UUID)

    @Throws(Exception::class)
    fun updatePassword(updatePasswordDTO: UpdatePasswordDTO)

    @Throws(DataNotFoundException::class)
    fun softDeleteUser(userId: UUID)

    @Throws(DataNotFoundException::class)
    fun restoreUser(userId: UUID)

    fun validateAccountBalance(user: User, basePrice: Long)

    fun existsByEmail(email: String): Boolean

    fun existsByPhoneNumber(phoneNumber: String): Boolean

    fun existsById(id: UUID): Boolean

    fun saveUser(user: User)
    fun disconnect(user: User)
    fun findConnectedUsers(): List<User>
}
